### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by dark0ghost**

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -5,23 +5,25 @@ use teloxide::Bot;
 use std::collections::HashMap;
 use std::sync::{Arc, Mutex};
 
+static FORMAT_TEXT: &str = "Оберните код в теги: 3 символа ` до и после кода \
```
Comment (zamazan4ik) [inline src/utils.rs]: Since these messages are related to specific functions (and then should have smaller scopes) and there is no performance difference between in and out variants - please move texts back to the functions.

File: src/main.rs [hunk 2]
```diff
@@ -31,7 +41,7 @@ async fn run() {
         .parse::<u8>()
         .expect("Cannot convert THRESHOLD to u8");
 
-    let bot = Bot::from_env();
+    let bot: Bot = Bot::from_env();
```
Comment (zamazan4ik) [inline src/main.rs]: Remove this change since there is no reason to declare a type here

File: src/detection.rs [hunk 3]
```diff
@@ -3,6 +3,88 @@ use once_cell_regex::regex;
...
+    "typeid",
+    "virtual",
+    "module",
+    "final",
+    "override",
+    "float",
+    "double",
+    "void",
+    "vector",
+    "std::",
```
Comment (zamazan4ik) [inline src/detection.rs]: Remove `std::`. This change can lead to too high false positive rate without changing exsiting keywords and/or thresholds

File: src/detection.rs [hunk 4]
```diff
@@ -3,6 +3,88 @@ use once_cell_regex::regex;
 use regex::Regex;
 use teloxide::types::*;
 
+const KEYWORDS: [&str;77] = [
```
Comment (zamazan4ik) [inline src/detection.rs]: Move Keywords back to the function body as a static variable

File: src/commands.rs [hunk 5]
```diff
@@ -1,5 +1,12 @@
 use teloxide::{prelude::*, utils::command::BotCommand};
 
+static HELP_TEXT: &str = "Для форматирования однострочной конструкции используйте\
```
Comment (zamazan4ik) [inline src/commands.rs]: Another one. Revert this change.

File: Cargo.toml [hunk 6]
```diff
@@ -18,6 +18,7 @@ regex = "1.4.2"
 serde_json = "1.0.59"
 teloxide = "0.3.3"
 tokio = { version = "0.2.22", features = ["rt-threaded", "macros"] }
+openssl = { version = "0.10", features = ["vendored"] }
```
Comment (zamazan4ik) [inline Cargo.toml]: Why do we need this here? Another problem is that `vendored` version will not detect systems's root certificates. Also possibly some changes to the build process are required since openssl build requires C compiler, perl and other stuff (need to check it)

File: src/detection.rs [hunk 7]
```diff
@@ -15,99 +97,13 @@ pub fn maybe_formatted(maybe_entities: Option<&[MessageEntity]>) -> bool {
             _ => (),
         }
     }
-
-    return false;
+     false
```
Comment (zamazan4ik) [inline src/detection.rs]: Remove extra space

File: src/main.rs [hunk 8]
```diff
@@ -10,6 +10,16 @@ use std::collections::HashMap;
 use std::env;
 use std::sync::{Arc, Mutex};
 
+#[cfg(test)]
+mod tests {
+    use crate::detection::is_code_detected;
+
+    #[test]
+    fn is_code() {
+        assert!(is_code_detected("int main(){std::cout<<hello<<'\n'; return 0;}", 3));
```
Comment (zamazan4ik) [inline src/main.rs]: Since `std::` will be removed from patterns, this test should be changed too (lower threshold or add more keywords). And yeah - this code isn't valid C++ code :) `hello` -> `"hello"` at least

**Review by zamazan4ik — CHANGES_REQUESTED** [review]
Nice work! Especially thanks for the CI. However, small changes are required.

File: Cargo.toml [hunk 9]
```diff
@@ -18,7 +18,7 @@ regex = "1.4.2"
 serde_json = "1.0.59"
 teloxide = "0.3.3"
 tokio = { version = "0.2.22", features = ["rt-threaded", "macros"] }
-openssl = { version = "0.10", features = ["vendored"] }
+
```
Comment (zamazan4ik) [inline Cargo.toml]: Remove redundant newline

File: src/main.rs [hunk 10]
```diff
@@ -41,7 +41,7 @@ async fn run() {
         .parse::<u8>()
         .expect("Cannot convert THRESHOLD to u8");
 
-    let bot: Bot = Bot::from_env();
+    let bot= Bot::from_env();
```
Comment (zamazan4ik) [inline src/main.rs]: Missed space before `=`

File: src/detection.rs [hunk 11]
```diff
@@ -3,7 +3,28 @@ use once_cell_regex::regex;
 use regex::Regex;
 use teloxide::types::*;
 
-const KEYWORDS: [&str;77] = [
+const
```
Comment (zamazan4ik) [inline src/detection.rs]: Remove extra word

File: src/detection.rs [hunk 12]
```diff
@@ -3,7 +3,28 @@ use once_cell_regex::regex;
...
+
+static INSTANCE: OnceCell<[&'static str; 76]> = OnceCell::new();
```
Comment (zamazan4ik) [inline src/detection.rs]: Move instance to the function too.

File: src/detection.rs [hunk 13]
```diff
@@ -3,7 +3,28 @@ use once_cell_regex::regex;
...
+        None => return false,
+    };
+
+    for entity in entities.iter() {
+        match entity.kind {
+            MessageEntityKind::Code | MessageEntityKind::Pre { .. } => return true,
+            _ => (),
+        }
+    }
+     false
```
Comment (zamazan4ik) [inline src/detection.rs]: Remove extra space before `false`

File: src/detection.rs [hunk 14]
```diff
@@ -3,7 +3,28 @@ use once_cell_regex::regex;
...
+}
+
+pub fn is_code_detected(text: &str, threshold: u8) -> bool {
+    let re: &Regex = regex!(INSTANCE
+        .get_or_init(||[
     "namespace",
```
Comment (zamazan4ik) [inline src/detection.rs]: Wrong formatting. Revert formatting as it was done before.

**Review by zamazan4ik — CHANGES_REQUESTED** [review]

**Comment by dark0ghost** [comment]
after looking wenhook.rs it seems to me unsafe to bind a webhook to a token. I suggest using https://github.com/nikolay-govorov/nanoid

**Comment by zamazan4ik** [comment]
Cannot really understand, why it's unsafe to use a token there (since token is unknown for anyone except Telegram servers).

**Comment by dark0ghost** [comment]
there are applications that check for available paths and also if you forget to specify in the robot.txt then the path can be indexed in search engines

**Comment by zamazan4ik** [comment]
There is so smaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaal chance, that Telegram Bot token will be bruteforced (if it's possible - Telegram has a really huge security hole). How it can be indexed in `robot.txt` - idk, since I see no way, how search engine will get my private Telegram Bot token.

**Comment by dark0ghost** [comment]
I think that it would be better to change the way of getting the link

**Comment by zamazan4ik** [comment]
Maybe you are right but not for security reasons since current approach is secure and there is no (at least I don't see) any way without hacking Telegram servers (or user's storage with Telegram bot token) to get the token.

However the possible improvement can be like: provide webhook bind url with environment variable. It's highly configurable to end-user way to set any uri, as they wish.

If you want to do such change, feel free to do it in another PR.

**Comment by dark0ghost** [comment]
ok

**Comment by zamazan4ik** [comment]
So for now fix the last warn and seems like it can be merged :)

**Comment by dark0ghost** [comment]
what? I fixed everything

**Comment by zamazan4ik** [comment]
Unfortunately - no. See detection.rs, line with `static INSTANCE: OnceCell<[&'static str; 76]> = OnceCell::new();`. It shall be moved back to the function as well.

**Review by zamazan4ik — APPROVED** [review]
Thank you for the work! Approved.
